_posts/2020-06-30-sparklyr-1.3.0-released/sparklyr-1.3.0.Rmd

```diff
@@ -0,0 +1,195 @@
+
+
+## Higher-order Functions
+
+[Higher order functions](https://issues.apache.org/jira/browse/SPARK-19480) are built-in Spark SQL constructs that allow user-defined lambda expressions to be applied efficiently to complex data types such as arrays and structs. As a quick demo to see why higher-order functions are useful, let's say one day Scrooge Mcduck dived into his huge vault of money and found large quantities of pennies, nickels, dimes, and quarters. Having an impeccable taste in data structure, he decided to store the quantities and face values of everything into 2 Spark SQL array columns:
+
+```{r echo=TRUE, eval=FALSE}
+library(sparklyr)
+
+sc <- spark_connect(master = "local", version = "2.4")
+coins_tbl <- copy_to(
+  sc,
+  tibble::tibble(
+    quantities = list(c(4000, 3000, 2000, 1000)),
+    values = list(c(1, 5, 10, 25))
+  )
+)
+```
+
+thus declaring his net worth of 4k pennies, 3k nickels, 2k dimes, and 1k quarters. To help Scrooge Mcduck calculate the total value of each type of coin in Spark 2.4 or higher, we can apply `hof_zip_with`, the sparklyr equivalent of [ZIP_WITH](https://spark.apache.org/docs/latest/api/sql/index.html#zip_with), to `quantities` column and `values` column, combining pairs of elements from arrays in both columns. As you might have guessed, we also need to specify how to combine those elements, and what better way to accomplish that than a concise one-sided formula `~ .x * .y` in R, which says we want (quantity * value) for each type of coin? So, we have the following:
+
+```{r echo=TRUE, eval=FALSE}
+result_tbl <- coins_tbl %>%
+  hof_zip_with(~ .x * .y, dest_col = total_values) %>%
+  dplyr::select(total_values)
+
+result <- collect(result_tbl)
+result$total_values
+```
+```
+[[1]]
+[1]  4000 15000 20000 25000
+```
+
+with the result `4000 15000 20000 25000` telling us there are in total 40 dollars worth of pennies, 150 dollars worth of nickels, 200 dollars worth of dimes, and 250 dollars worth of quarters, as expected.
+
+Using another sparklyr function named `hof_aggregate`, which performs an [AGGREGATE](https://spark.apache.org/docs/latest/api/sql/index.html#aggregate) operation in Spark, we can then compute the net worth of Scrooge Mcduck based on `result_tbl`, storing the result in a new column named `total`:
+
+```{r echo=TRUE, eval=FALSE}
+net_worth_tbl <- result_tbl %>%
+  dplyr::mutate(zero = dplyr::sql("CAST (0 AS BIGINT)")) %>%
+  # ^^ because the starting value for our aggregate operation must have the correct
+  #    data type of BIGINT
+  hof_aggregate(start = zero, ~ .x + .y, expr = total_values, dest_col = total) %>%
+  dplyr::select(total)
+
+net_worth <- collect(net_worth_tbl)
+net_worth$total
```

Comment:
I think we can also use `pull` here:

```suggestion
net_worth_tbl <- result_tbl %>%
  dplyr::mutate(zero = dplyr::sql("CAST (0 AS BIGINT)")) %>%
  # ^^ because the starting value for our aggregate operation must have the correct
  #    data type of BIGINT
  hof_aggregate(start = zero, ~ .x + .y, expr = total_values, dest_col = total) %>%
  dplyr::select(total) %>%
  dplyr::pull(total)
```